### PR TITLE
Add CoreML-stable RMSNorm for llama eager paths (#19523) (#19523)

### PR DESCRIPTION
**Pull request by telgamal-1**
Summary:
The standard `RMSNorm` formulation `x * rsqrt(mean(x²)) * weight` is numerically unstable on CoreML/ANE because the explicit FP32 cast around the mean reduction is silently stripped from the lowered graph, leaving the squared sum to overflow in FP16. The ANE PTE then diverges from the eager reference even on checkpoints fine-tuned in BF16/FP16.

This diff introduces `RMSNormCoreML` in `examples/models/llama/norm.py`. The module expresses the normalization as `x * sqrt(d) / vector_norm(x, dim=-1)` — `torch.linalg.vector_norm` keeps the reduction in a single op that survives CoreML lowering, so FP16 inference remains stable.

To avoid `0 / 0 = NaN` on zero-padded positions (chunked prefill in `StaticAttentionIOManager` pads each chunk to `input_len` with zeros), the denominator is floored with `sqrt(dim * eps)`. This matches standard RMSNorm's `rsqrt(mean(x²) + eps)` semantics on a zero input and is large enough to survive fp16 — a plain `1e-6` underflows. Real (non-zero) tokens satisfy `vector_norm(x) >> sqrt(dim * eps)`, so the floor is a no-op on real positions.

A new `use_coreml_norm: bool = False` field on `ModelArgs` opts into the new norm without disturbing existing models. When True, every llama-side norm site constructs `RMSNormCoreML`:
- `llama_transformer.py`: `attention_norm`, `ffn_norm`, the final `self.norm` on `Transformer`.
- `attention.py`: `q_norm_fn` / `k_norm_fn` in the affine QK-norm path, AND the `else` branch of `_init_qk_norms` (the scaleless / non-affine QK-norm path that the original landing missed).
- `static_attention.py`: `q_norm` / `k_norm` in the scaleless path, propagated through `from_attention_mha` by detecting `rms_norm_class is RMSNormCoreML`.

The QNN/HTP export path is untouched and continues to use `torch.nn.RMSNorm`.

Reviewed By: YIWENX14

Differential Revision: D104862210

Pulled By: telgamal-1

### CODE REVIEW

File: examples/models/llama/norm.py [hunk 1]
```diff
@@ -57,6 +57,47 @@ def __init__(self, dim: int, eps: float = 1e-6):
...
+        """
+        CoreML-friendly RMSNorm — uses `torch.linalg.vector_norm` so the op is
+        preserved in the CoreML graph for numerical stability.
+
+        Args:
+            dim (int): The dimension of the input tensor.
+            eps (float, optional): Stored for API compatibility; ignored in the math.
+
+        Attributes:
+            eps (float): Stored for API compatibility; not consumed by `_norm`.
```
Comment (metascroy) [inline examples/models/llama/norm.py]: Can we assert eps is 0 rather than silently drop it?

Comment (telgamal-1) [reply examples/models/llama/norm.py]: Added an explicit assert of eps is set to 0

File: examples/models/llama/norm.py [hunk 2]
```diff
@@ -57,6 +57,51 @@ def __init__(self, dim: int, eps: float = 1e-6):
         self.weight.requires_grad = False
 
 
+class RMSNormCoreML(torch.nn.Module):
```
Comment (metascroy) [inline examples/models/llama/norm.py:60]: How does this differ from:

https://github.com/pytorch/executorch/blob/7cd209d48d0827a1835fd0bfbe4d0758282281a9/examples/apple/coreml/llama/llama_transformer.py#L112

Can we consolidate? Putting it here is fine, but then import this version into examples/apple/coreml/llama/llama_transformer.py.

File: examples/models/llama/attention.py [hunk 3]
```diff
@@ -425,21 +426,29 @@ def _init_norms(self, args: ModelArgs) -> None:
...
         if self.use_qk_norm:
             if args.qk_norm_affine:
-                self.q_norm_fn = RMSNorm(
-                    self.head_dim,
-                    eps=args.norm_eps,
-                    add_unit_offset=args.rms_norm_add_unit_offset,
-                )
-                if self.has_kv_weights:
-                    self.k_norm_fn = RMSNorm(
+                if args.use_coreml_norm:
```
Comment (metascroy) [inline examples/models/llama/attention.py]: Does this have to be integrated so far down?

Could we not leave llama_transformer/static attention as is, introduce the new norm in norm.py, and then do a module swap from RMSNorm -> CoreMLRMsNorm at export time?

File: examples/models/llama/norm.py (hunk 2 above)
Comment (telgamal-1) [reply examples/models/llama/norm.py:60]: I imported the new version in examples/apple/coreml/llama/llama_transformer.py because it was tested to not produce NaN in QAT

File: examples/models/llama/attention.py (hunk 3 above)
Comment (telgamal-1) [reply examples/models/llama/attention.py]: addresed and now using the strategy to replace_rms_norm_for_coreml_

**Comment by metascroy** [comment]
LGTM! You need to run the lintrunner, though

**Review by metascroy — DISMISSED** [review]

**Review by metascroy — APPROVED** [review]

**Review by YIWENX14 — APPROVED** [review]
